infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/cache/jpa/JpaKubernetesRuntimeStateCache.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.cache.jpa;
+
+import com.google.inject.persist.Transactional;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.persistence.EntityExistsException;
+import javax.persistence.EntityManager;
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.core.db.jpa.DuplicateKeyException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.cache.KubernetesRuntimeStateCache;
+import org.eclipse.che.workspace.infrastructure.kubernetes.model.KubernetesRuntimeState;
+import org.eclipse.che.workspace.infrastructure.kubernetes.model.KubernetesRuntimeState.RuntimeId;
+
+/**
+ * JPA based implementation of {@link KubernetesRuntimeStateCache}.
+ *
+ * @author Sergii Leshchenko
+ */
+public class JpaKubernetesRuntimeStateCache implements KubernetesRuntimeStateCache {
+
+  private final Provider<EntityManager> managerProvider;
+
+  @Inject
+  public JpaKubernetesRuntimeStateCache(Provider<EntityManager> managerProvider) {
+    this.managerProvider = managerProvider;
+  }
+
+  @Override
+  public boolean putIfAbsent(KubernetesRuntimeState runtimeState) throws InfrastructureException {
+    try {
+      doPutIfAbsent(runtimeState);
+      return true;
+    } catch (DuplicateKeyException | EntityExistsException e) {
+      return false;
+    } catch (RuntimeException e) {
+      throw new InfrastructureException(e.getMessage(), e);
+    }
+  }
+
+  @Transactional(rollbackOn = {RuntimeException.class, InfrastructureException.class})
+  @Override
+  public Set<RuntimeIdentity> getIdentities() throws InfrastructureException {
+    try {
+      return managerProvider
+          .get()
+          .createQuery("SELECT r FROM KubernetesRuntime r", KubernetesRuntimeState.class)
```

Comment:
can we use named query here?

Reply:
Good catch. Replaced with a named query.